nemo/collections/asr/metrics/wer.py

```diff
@@ -146,48 +186,462 @@ def ctc_decoder_predictions_tensor(
             Either a list of str which represent the CTC decoded strings per sample,
             or a list of Hypothesis objects containing additional information.
         """
-        hypotheses = []
-        # Drop predictions to CPU
-        predictions = move_dimension_to_the_front(predictions, self.batch_dim_index)
-        prediction_cpu_tensor = predictions.long().cpu()
-        # iterate over batch
-        for ind in range(prediction_cpu_tensor.shape[0]):
-            if self.fold_consecutive:
-                prediction = prediction_cpu_tensor[ind].detach().numpy().tolist()
+        decoder_outputs = move_dimension_to_the_front(decoder_outputs, self.batch_dim_index)
```

Comment:
I have a concern that maybe it is not the best idea to do decoding inside the WER calculation method. Isn't it better to separate them? I mean first do the decoding, then pass the final decoded results to calculate the WER?
encoder-forward->decoder-forward->decoding->wer_calculation. Separating them may cause more difficulty I am not aware. Just something to consider.

Reply:
So I decided to seperate them in order to make them more modular. For example, without the decoding class, the joint would need to implement parts of decoding, but they are basic pytorch modules they should not implement complex logic like this. Without decoding, wer would need to deal with many cases, but it is a metric, it should not deal with decoding.

So decoding acts as intermediate - takes the output of any joint, decodes it with additional information (say beam search, or timestamps etc), and the output then goes into WER. 

Note that the class may exist inside wer.py, but it doesn't even know of the existence of wer.py - it doesn't interact with the metric at all. The metric instead knows of the decoding layer, and uses its public methods to calculate what it needs from the decoding strategy. So decoding class itself has no notion of metric calculation.